Sources/Apodini/Application/Application+Connection.swift

```diff
@@ -29,27 +30,68 @@ extension Application {
 /// Represents the state of a connection to the client.
 /// For non-client-streaming requests,
 /// the default state should be `.end`.
-public enum ConnectionState: String {
+public enum ConnectionState: String, Hashable {
     /// The request is part of a client stream,
     /// and there are more requests to follow.
     case open
     /// The request is the last in the current client stream.
     case end
+    /// The previous request was the last request in the stream,
+    /// and the client has now requested the stream be closed.
+    /// - Note: This state exists to handle cases where a connection is closed after the last request has been received.
+    ///         With some protocols (e.g. gRPC), it is not guaranteed that a client stream is always closed with a final request,
+    ///         but instead the stream can also be kept open for some time after sending the last requesst, with a client then
+    ///         eventually closing the stream by sending a request-less HTTP frame to the server.
+    ///         In such cases (i.e. a stream being closed w/out there being an accompanying final request), the handler is invoked
+    ///         once again, with this connection state.
+    ///         It is guaranteed that a handler will be invoked with only the `end` state or with only the `close` state.
+    ///         A hander will never be called for both of these states.
```

Comment:
There are some protocols (e.g. gRPC) where a client-side-streaming connection can be closed w/out an actual request/message. In this case, we still need to invoke the handler one last time to give it the opportunity to send a response.
Consider an example handler that collects names from a client, and then, when the client indicates that there are no further names, responds with a concatenated list of all these clients. In gRPC, the handler would receive each name with an `open` connection state, and the client would then send an empty HTTP frame indicating that the connection should be closed. The issue is that we only know retroactively that the most recent request was in fact the last one, so we need to somehow be able to tell the handler that it should now send its final response. This is what the close state is for (and also why a handler will only ever receive one of close and end, but never both.)

One additional thing (and this was @PSchmiedmayer's idea) is that this final `.close` handler invocation will receive the parameters from the most recent `.open` invocation. This means that the handler could, in theory, still access its parameters without crashing.

You can have a look at the playground test web service, there's an example of one such handler in there. In most cases the differentiation between close and end shouldn't matter too much, and can be handled pretty elegantly using a switch statement with a fallthrough

Reply:
Thanks for taking the time to explain this in a bit more detail. Helped a lot to understand the situation.

So, `close` and `end` are pretty much the same thing (from a Handler implementation perspective), with the difference, that with `close` the Handler isn't updated (e.g. Parameters hold the value from the previous request). 

Does this always hold? Is the Handler state always equal to that from the previous request when receiving a `close`?

As `close` and `end` pretty much model the same thing, with the difference that `close` is caused by an empty request, would it make sense to merge those two cases and encode this property via an associated value, which indicates if the request was empty or not? Note, this builds around my assumptions outlined above. But to me, this would maybe model the situation a bit better? And Handlers which do not care about it, don't need to expose those "edge case" in their implementation. Also Handlers that need that (like the `StreamingGreeter_CS` Handler in the example web service) could benefit from it, as a `if` syntax is more quickly recognized by a glance than `fallthrough`(?). Lastly, we wouldn't need the Equatable workaround.
But not sure if such an approach was already discussed and I'm missing something essential.